### PR TITLE
Fix incorrect calculation of onMouseEnter/Leave component path

### PR DESCRIPTION
**Pull request by gaearon**
Fixes https://github.com/facebook/react/issues/10906 and https://github.com/facebook/react/issues/11152.
We forgot to compare alternates so in some cases, the from/to path was calculated incorrectly.

See individual commits. The first one adds a test, the last one fixes the bug. The ones in the middle prepare the code to look readable after the bugfix.

I verified both https://github.com/facebook/react/issues/10906 and https://github.com/facebook/react/issues/11152 reproduced before, but no longer reproduce after the bugfix.

### CODE REVIEW

File: src/renderers/shared/shared/ReactTreeTraversal.js [hunk 1]
```diff
@@ -110,18 +110,38 @@ function traverseTwoPhase(inst, fn, arg) {
  * "entered" or "left" that element.
  */
 function traverseEnterLeave(from, to, fn, argFrom, argTo) {
-  var common = from && to ? getLowestCommonAncestor(from, to) : null;
-  var pathFrom = [];
-  while (from && from !== common) {
+  const common = from && to ? getLowestCommonAncestor(from, to) : null;
+  const pathFrom = [];
+  while (true) {
```
Comment (sebmarkbage) [inline src/renderers/shared/shared/ReactTreeTraversal.js]: We try to avoid `while (true)` since (at least in older versions) it has been known to deopt.

I appreciate the rewrite for the temp variable and easy of read. But arguably this just make the condition seem a lot more complicated than it really is.

Comment (gaearon) [reply src/renderers/shared/shared/ReactTreeTraversal.js]: Didn't it deopt when you have `continue`s rather than `break`? I don't remember. @trueadm

**Comment by gaearon** [comment]
Like this?

```js
  while (from && from !== common && (!from.alternate && from.alternate !== common)) {
```

I'm not sure how else to write it in a single line. I’m finding this really hard to follow, especially because the `null` check (since `common` can be null) complicates the condition.

**Comment by gaearon** [comment]
Or, rather,

```js
while (from && from !== common && (!from.alternate || from.alternate !== common)) {
```

See, I already made a mistake there. 😛

**Comment by gaearon** [comment]
Do you like this?

```js
  while (from && (!common || (from !== common && from.alternate !== common))) {
```

If we check for `common` first then we shouldn't need to check `alternate` there.

**Comment by gaearon** [comment]
Meh. I find all of these hard to follow and I don't know if they're correct.
If you feel strongly let me know what style you prefer and I'll rewrite in a separate PR.
